gto-support-kotlin-coroutines/src/main/java/org/ccci/gto/android/common/kotlin/coroutines/ReadWriteMutex.kt

```diff
@@ -0,0 +1,43 @@
+package org.ccci.gto.android.common.kotlin.coroutines
+
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+interface ReadWriteMutex {
+    val write: Mutex
+    val read: Mutex
+}
+
+fun ReadWriteMutex(): ReadWriteMutex = ReadWriteMutexImpl()
+
+private class ReadWriteMutexImpl : ReadWriteMutex {
+    private val stateMutex = Mutex()
+    private var readers = 0
+
+    override val write = Mutex()
+    override val read = object : Mutex {
+        override suspend fun lock(owner: Any?) {
+            stateMutex.withLock(owner) {
```

Comment:
Certainly not a biggie, but I'm wondering if passing `owner` here is correct. It means reader coroutines can't be reentrant. Which maybe is fine, but it's somewhat of an artificial limitation, since reader reentrancy wouldn't cause deadlocks. (From what I gather on google, 'owner' is intended to help you catch deadlocks pre-emptively, as long as you pass a coroutine-unique owner.)

Speaking of owners & deadlocks, if a coroutine acquires a read lock and then tries to acquire a write lock, the current 'owner' code will prevent a deadlock. But, that's only true if there's not another reader; if a second coroutine first acquires a read lock, and then the first coroutine acquires a read lock and then tries to acquire a write lock -- the current 'owner' code will won't trigger an exception, but will deadlock. Not sure how hard this would be to fix, but if you don't you probably should punt and document that this class doesn't fully implement the 'owner' safeguarding stuff. That's probably what I'd do.

Reply:
coroutines Mutex's aren't actually reentrant, so that's not a big deal. the `stateMutex` here only holds the lock until the end of the block.

There is a potential stall if `read.unlock()` is called while `read.lock()` is trying to get the first read lock. But assuming proper usage of the Mutex this state should never be reached. `read.unlock()` should only be called when readers > 0, at which point `read.lock()` won't potentially block on `write.lock()`.

I will put a couple of limit checks in place to error out if this invalid state is ever reached.

Reply:
from documentation of the `owner` parameter:
>     * @param owner Optional owner token for debugging. When `owner` is specified (non-null value) and this mutex
>     *        is already locked with the same token (same identity), this function throws [IllegalStateException].

Reply:
I do need to pass something else to `write.lock()` and `write.unlock()` for the owner though, if a different reader locks than unlocks, it will throw an `IllegalStateException`

Reply:
alright, I think I fully understand everything packed in here now, I effectively disabled the owner functionality for the shared read lock and added a documentation note about it.